### PR TITLE
Allow RetryLoop to be type safe

### PR DESCRIPTION
**Pull request by torcolvin**
- Use generics to allow use of type safe RetryLoops
- Drop need in tests to always assert on errr != nil for some rest utility test functions

## Pre-review checklist
- [x] Removed debug logging (`fmt.Print`, `log.Print`, ...)
- [x] Logging sensitive data? Make sure it's tagged (e.g. `base.UD(docID)`, `base.MD(dbName)`)
- [x] Updated relevant information in the API specifications (such as endpoint descriptions, schemas, ...) in `docs/api`

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGateway-Integration/build?delay=0sec)
- [x] `GSI=true,xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration/3096/ (known independent flake)

### CODE REVIEW

File: rest/utilities_testing.go [hunk 1]
```diff
@@ -2404,12 +2383,12 @@ func WaitAndAssertBackgroundManagerState(t testing.TB, expected db.BackgroundPro
 
 func WaitAndAssertBackgroundManagerExpiredHeartbeat(t testing.TB, bm *db.BackgroundManager) bool {
```
Comment (gregns1) [inline rest/utilities_testing.go:2384]: This doesn't look like its used at all anywhere, should we just remove it?

**Review by gregns1 — APPROVED** [review]
LGTM - can take or leave the comment I have made

File: rest/utilities_testing.go (hunk 1 above)
Comment (torcolvin) [reply rest/utilities_testing.go:2384]: I'll remove in a different PR, I have one that wipes out all the public functions that are unused using https://pkg.go.dev/golang.org/x/tools/cmd/deadcode

Since Sync Gateway isn't a library, we can easily remove dead public functions.
